File: src/data/courses/dsa/binary-search-trees/views-vertical-problems/vertical-order/code/java/VerticalOrder.java
import java.util.*;
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        if (root==null) return new ArrayList<>();
        TreeMap<Integer,List<Integer>> map = new TreeMap<>();
        Queue<Object[]> q = new LinkedList<>(); q.add(new Object[]{root,0});
        while (!q.isEmpty()) { Object[] c=q.poll(); TreeNode n=(TreeNode)c[0]; int hd=(int)c[1];
            map.computeIfAbsent(hd,k->new ArrayList<>()).add(n.val);
            if(n.left!=null) q.add(new Object[]{n.left,hd-1});
            if(n.right!=null) q.add(new Object[]{n.right,hd+1}); }
        return new ArrayList<>(map.values());
    }
}

File: src/data/courses/dsa/binary-search-trees/views-vertical-problems/vertical-order/code/python/vertical_order.py
from collections import deque, defaultdict
class Solution:
    def verticalOrder(self, root):
        if not root: return []
        cols = defaultdict(list)
        queue = deque([(root, 0)])
        while queue:
            node, hd = queue.popleft()
            cols[hd].append(node.val)
            if node.left: queue.append((node.left, hd-1))
            if node.right: queue.append((node.right, hd+1))
        return [cols[k] for k in sorted(cols)]
